README.md
# A WYSIWYG-texteditor based on pygame for pygame

![PyPI](https://img.shields.io/pypi/v/pygame-texteditor?color=%233775A9&label=pypi%20package&style=plastic)
![GitHub](https://img.shields.io/github/license/CribberSix/pygame-texteditor?style=plastic)

## Introduction & examples

The text editor can be inserted into any existing pygame window.
A minimal code example of it being activated within an existing pygame window can be found below.

The code editor comes with line numbers and syntax highlighting for python if enabled:

![](./resources/example_one.png)

Example with default configuration:

![](./resources/example_two.png)

## Usage

The texteditor takes 5 obligatory parameters and 4 optional parameters.

##### Obligatory parameters
- ```offset_X``` : integer - the offset from the left border of the pygame screen
- ```offset_y``` : integer - the offset from the top border of the pygame screen
- ```editor_width``` : integer - the width of texteditor
- ```editor_height``` : integer - the height of texteditor
- ```screen``` : pygame display surface - on which the texteditor is to be displayed

##### Optional Parameters with default values

- ```display_line_numbers``` - a boolean enabling showing line numbers
    > Default: ```False```
- ```style``` - a String setting the color scheme of editor and syntax highlighting
    > Default: ```'dark'```
- ```syntax_highlighting_python``` - a boolean enabling syntax highlighting for Python code
    > Default: ```False```
- ```font_size``` - an integer to set for the font size.
    > Default: ```16```

## Setup and configuration

##### Minimal texteditor setup example

```python
import pygame
from pygame_texteditor import TextEditor

# minimal pygame setup
pygame.init()
screen = pygame.display.set_mode((500, 600))
pygame.display.set_caption("Pygame")
pygame.display.get_surface().fill((200, 200, 200))  # background coloring

# Instantiation & customization of the text editor
TX = TextEditor(
    offset_x=50, offset_y=50, editor_width=500, editor_height=400, screen=pygame.display.get_surface()
)
TX.set_line_numbers(True)
TX.set_syntax_highlighting(True)
TX.set_font_size(18)

# TextEditor in the pygame-loop
while True:
    # INPUT - Mouse + Keyboard
    pygame_events = pygame.event.get()
    pressed_keys = pygame.key.get_pressed()
    mouse_x, mouse_y = pygame.mouse.get_pos()
    mouse_pressed = pygame.mouse.get_pressed()

    # displays editor functionality once per loop
    TX.display_editor(pygame_events, pressed_keys, mouse_x, mouse_y, mouse_pressed)
    pygame.display.flip()  # updates pygame window

```

##### Retrieving text from the editor

The editor offers the function `get_text_as_string()` to retrieve the entire text
as a String from the editor. Lines are separated by the new line character ```\n```.

The editor offers the function `get_text_as_list()` to retrieve the entire text as a list from the editor.
Each String-item in the list represents one line from the editor.

##### Removing text from the editor

The editor offers the function `clear_text()` to clear the editor of any text.

##### Inserting text into the editor

Inserting text can be done by using one of the two available functions:
1. With a list of strings in which each string represents one line, or
2. With a string which includes linebreak characters which get parsed.

```
set_text_from_list(["First line", "Second Line.", "Third Line."]
set_text_from_string("First line.\nSecond line.\nThird Line")
```

## Customization

#### Cursor mode

Cursor mode can either be `static` or `blinking` (=default).

```python
TX = TextEditor(...)
TX.set_cursor_mode("static")
TX.set_cursor_mode("blinking")
```

#### Key repetition speeds

While a key is being held, multiple key events are being triggered.
The delay of the first repetition as well as the interval between all sequential key triggers can be
customized by using the function `set_key_repetition(delay=300, intervall=30)`.

From the [official documentation](http://www.pygame.org/docs/ref/key.html#pygame.key.set_repeat):
> The delay parameter is the number of milliseconds before the first repeated pygame.KEYDOWN event will be sent.
> After that, another pygame.KEYDOWN event will be sent every interval milliseconds.


#### Font Customization

The editor uses a ttf file to set the font for the editor. By default, the Courier monospace font is used.

A custom font can be loaded with the following method, passing an *absolute* path:
- `set_font_from_ttf("X:\path\to\custom\font.ttf")`

DISCLAIMER: As the width of a letter (space) is only calculated once after setting the font_size, any fonts that are not monospace will lead to the editor not working correctly anymore, as it cannot be determined correctly between which letters the user clicked.

#### Font size

Font size can be customized with the command `set_font_size(size)` - the parameter is an integer
with the default value `16` to be able to reset it.

#### Line Numbers
Line numbers can be shown on the left side of the editor. Line numbers begin with 0 as is the Pythonian way.

Line numbers can be enabled and disabled with ```set_line_numbers(Boolean)```.


#### Syntax Highlighting

The editor comes with syntax highlighting for Python code. Tokenization is based on the ```pygment``` package.

Syntax highlighting can be enabled/disabled with ```set_syntax_coloring(boolean_value)```.

The syntax colors being used are also specified in the yml style file.


#### Color-scheme customization

The editor uses a yml file to set the color-scheme for the editor itself and for the syntax coloring.

Two styles are delivered with the editor, they can be activated respectively by:
- `set_colorscheme("dark")`
- `set_colorscheme("bright")`

A custom style can be loaded with the following method from a created yml file:
- `set_colorscheme_from_yaml("X:\path\to\custom\filename.yml")`

All keys must be present with values. Acceptable values are
RGB colors in the following format: ```(255, 255, 255)``` or ```255, 255, 255```.

The following keys are required in the ```stylename.yml``` file, syntax colors are only used if syntax
highlighting is enabled, but are still required to be included.

**Editor colors** (source: bright.yml)

- `codingBackgroundColor: (255, 255, 255)`
- `codingScrollBarBackgroundColor: (49, 50, 50)`
- `lineNumberColor: (255, 255, 255)`
- `lineNumberBackgroundColor: (60, 61, 61)`
- `textColor: (255, 255, 255)`

**Syntax colors** (source: bright.yml)

- `textColor_normal: (0, 255, 255)`
- `textColor_comments: (119, 115, 115)`
- `textColor_quotes: (227, 215, 115)`
- `textColor_operators: (237, 36, 36)`
- `textColor_keywords: (237, 36, 36)`
- `textColor_function: (50, 150, 36)`
- `textColor_builtin: (50, 50, 136)`


CHANGELOG.md
# Changelog

### 0.7.2: Added parameter to usage of pygment lexer

Added the parameter to `PythonLexer(ensurenl=False)` in order to avoid the issue from
[pygment](https://github.com/pygments/pygments/issues/610) as identified by [brno32](https://github.com/brno32)!


### 0.7.1: Bugfix for displaying line numbers correctly and doc-update

- Minor fix concerning a bug which caused the line number background to have a wrong width when first initialized.
- Improved documentation concerning README.md and [how-to-use.py](https://github.com/CribberSix/pygame-texteditor/blob/master/how-to-use.py).

### 0.7.0: QoL update - BREAKING CHANGES

- A lot of internal variables but also some parameters have been renamed to enable easier maintenance of the codebase. This might break existing implementations using the package if you hardcoded parameter names upon function calls or if you called internal variables.
- Removed unnecessary class attributes.
- Fixed a visual bug which caused the end of th escrollbar to be displayed outside the area of the texteditor.
- Fixed an issue [#11](https://github.com/CribberSix/pygame-texteditor/issues/11) which caused line numbers to be rendered incorrectly if a large font size was set.
- Fixed an isse [#12](https://github.com/CribberSix/pygame-texteditor/issues/10) which caused some lines to be un-selectable when a large font was set.
- Fixed an issue which caused the editor to crash if you used the arrow keys to scroll out of the visible lines.
- Fixed an issue which caused line numbers to extend over their background and even outside the editor if numbers went above two digits.

### 0.6.8: Font customization

Added option to use a custom (monospace) font, thanks to [brno32](https://github.com/brno32)'s contribution!

### 0.6.7: Cursor customization

Added option to have a static (or blinking) cursor.

### 0.6.5: Pygame version

Changed requirements concerning pygame to `>= 1.9.6` as version `2.0.1` has backwards compatibility.

### 0.6.4: Bugfix

Fixed highlight-rendering (adjusted to actual line height + gap!)

### 0.6.3: Font size customization

Implemented new function to customize font-size.


Environment.py
import pygame, sys
from io import StringIO
from Scripts.LearningEnvironment.TextEditor import TextEditorClass
from Scripts.MiscScripts import Button
from Scripts.LearningEnvironment.Curriculum import Lesson

class LearningEnvironment_Class():


    def __init__(self):
        self.idea_icon = pygame.image.load("Graphics/LearningEnvironment/Idea.png").convert()
        self.checkCodeGreen_image = pygame.image.load("Graphics/LearningEnvironment/CheckCodeGreen.png").convert()
        self.checkCodeRed_image = pygame.image.load("Graphics/LearningEnvironment/CheckCodeRed.png").convert()
        self.checkCodes_image = pygame.image.load("Graphics/LearningEnvironment/CodeChecks.png").convert()

        self.HeadlineFont = pygame.font.Font("Misc/Ultra-Regular.ttf",20)
        self.Courier_Text_15_Font = pygame.font.Font("Misc/Courier.ttf", 15)
        self.Ubuntu_Regular_25 = pygame.font.Font("Misc/UbuntuMono-Regular.ttf", 25)
        self.Ubuntu_Bold_25 = pygame.font.Font("Misc/UbuntuMono-Bold.ttf", 25)
        self.SourceCodePro_Light_25 = pygame.font.Font("Misc/SourceCodePro-Light.ttf", 25)
        self.SourceCodePro_Light_15 = pygame.font.Font("Misc/SourceCodePro-Light.ttf", 15)

        self.Headline_Text = self.SourceCodePro_Light_25.render( "Learn" ,1,(10,10,10))

        self.buttonCreater = Button.Buttons()
        self.output_string = ""
        self.backgroundColor = (21, 43, 57)
        self.instructionBackgroundColor = (250,250,250)
        self.headlineColor = (212, 215, 214)
        self.codingOutputBackgroundColor = (70, 88, 117)
        self.codingOutputLetterColor = (255,255,255)

        self.codeCheckString = ""

        # FPS
        self.FPS = 30
        self.clock = pygame.time.Clock()

        self.CheckCodeComplete = 0
        self.CheckOutputComplete = 0
        self.CheckCodeComplete_Bool = False
        self.lessonsfinishedduringthissession = 0

    def display_Learner(self, screen, SettingsObj, lessonNumber):
        self.SettingsObj = SettingsObj

        self.w, self.h = pygame.display.get_surface().get_size()
        self.TX = TextEditorClass(self.w*0.3, 0, int((self.w*0.7)), self.h*0.7, screen) # X_offset, Y_offset, codeAreaWidth, codeAreaHeight   # TODO: (check how high I actually want it to be!)
        self.Lesson = Lesson.LessonClass(screen)
        self.Lesson.loadLesson(lessonNumber, self.w * 0.3 )


        while(True):
            ####_________ENVIRONMENTS_________####
            pygame.draw.rect(screen, self.codingOutputBackgroundColor, (self.w*0.3, self.h*0.7, self.w*0.7, self.h*0.3))
            pygame.draw.rect(screen, self.instructionBackgroundColor, (0, 0, self.w*0.3, self.h)) # Instructions (left side - entire window)

            ####_________INSTRUCTIONS__________####
            # LEARN Icon+Caption
            pygame.draw.rect(screen, self.headlineColor, (0, 0, self.w*0.3, 40))
            screen.blit(self.idea_icon, (5, 0))
            screen.blit(self.Headline_Text, (50, 2))

            ####_________CURRICULUM_LESSON__________####
            # Contains all introductions, an example, the "Instruction-Header + Icon" and instructions
            self.Lesson.display()  # line-length + line-amount + header&icon are dynamically regulated.

            ####_________TEXTBOX__________####
            self.TX.display_CodeEditor()

            ####_________BUTTONS__________####
            # Run Code # TODO maybe delete this button - only have "check code" button.
            #if ("RunCode" == self.buttonCreater.createCodeEnvButton(screen, self.w*0.3, (self.h*0.7)+1, "RunCode" )):
            #    code_string = ""
            #    for line in self.TX.line_String_array:
            #        if line != "":
            #            code_string = code_string + "\n" + line
            #    oo = self.execute_Code(code_string)
            #    if oo != None:
            #        self.output_string = oo[1]
            #    else:
            #        self.output_string = ""

            # Run & Check Code.
            checkmyCode = self.buttonCreater.createCodeEnvButton(screen, self.w*0.3, (self.h*0.7)+1, "CheckCode" )
            if ("CheckCode" == checkmyCode):
                self.CheckCodeComplete = 0
                self.CheckOutputComplete = 0
                code_string = ""
                for line in self.TX.line_String_array:
                    if line != "":
                        code_string = code_string + "\n" + line
                oo = self.execute_Code(code_string)
                if oo != None:
                    self.output_string = oo[1]
                    if oo[0]:  # Code is executable (No Error occurred)

                        # Check user code for compliance with instructions & lesson parameters
                        # Code Check 1
                        if self.Lesson.CodeChecks > 0:
                            self.codeCheckString = code_string + "\n" + self.Lesson.Code_Check_1
                            t1 = self.execute_Code(self.codeCheckString)
                            if t1[0]:
                                self.CheckCodeComplete += 1
                            else:
                                print("Error during First Code Check.")
                                # TODO - Explain what / why? (Add "2nd Test not yet satisfied"?)

                        # Code Check 2
                        if self.Lesson.CodeChecks > 1:
                            self.codeCheckString = code_string + "\n" + self.Lesson.Code_Check_2
                            t2 = self.execute_Code(self.codeCheckString)
                            if t2[0]:
                                self.CheckCodeComplete += 1
                            else:
                                print("Error during Second Code Check.")
                                # TODO - Explain what / why? (Add "2nd Test not yet satisfied"?)


                        # Output Check 1
                        if self.Lesson.OutputChecks > 0:
                            if self.Lesson.Output_Check_1 in self.output_string:
                                self.CheckOutputComplete += 1
                            else:
                                print("Error during first output check.")
                                # TODO - Explain what / why? (Add "2nd Test not yet satisfied"?)


                        if self.CheckCodeComplete == self.Lesson.CodeChecks and self.CheckOutputComplete == self.Lesson.OutputChecks:
                            self.CheckCodeComplete_Bool = True  # enable continuing to next chapter / back to the game.

                else:
                    # Something went wrong.
                    # Code was not executable.
                    self.output_string = ""

            # Code Check Visuals
            screen.blit(self.checkCodes_image, (self.w*0.7, (self.h*0.7)+1))
            codeCheck_CoordX = self.w*0.7+150
            posChecks = self.CheckCodeComplete + self.CheckOutputComplete
            for _ in range(self.Lesson.CodeChecks + self.Lesson.OutputChecks):
                if posChecks > 0:
                    screen.blit(self.checkCodeGreen_image, (codeCheck_CoordX, (self.h*0.7)+1))
                    codeCheck_CoordX += 35
                    posChecks -= 1
                else:
                    screen.blit(self.checkCodeRed_image, (codeCheck_CoordX, (self.h*0.7)+1))
                    codeCheck_CoordX += 35

            # Continue to next Lesson Button
            if self.CheckCodeComplete_Bool:
                if "Continue" == self.buttonCreater.createCodeEnvButton(screen, self.w*0.3+300, (self.h*0.7)+1, "Continue" ):
                    self.CheckCodeComplete = 0 # Restore
                    self.CheckOutputComplete = 0 # Restore
                    self.CheckCodeComplete_Bool = False # Restore
                    self.output_string = "" # Restore
                    if self.Lesson.Continue:  # Continue to learn with the next lesson.
                        lessonNumber = self.display_Learner(screen, SettingsObj, lessonNumber+1) # command returns here.
                        return lessonNumber  # Return until the "stack of sub-lessons" are handled.
                    else:
                        return lessonNumber  # Return immediately to the game.

            # DEV Button # TODO: Remove
            #if "ToGame" == self.buttonCreater.createMenuButton(screen, "BackToGame", screen_wdith*0.5, screen_height*0.9, "ToGame"):
            #    return 1 # lessonNumber = 1; DEV Default

            ###_________CONSOLE__________####
            self.Console_Output_Text =  self.Courier_Text_15_Font.render(">>> Terminal <<<", 1, self.codingOutputLetterColor)
            screen.blit(self.Console_Output_Text, ((self.w * 0.3) + 10, (self.h * 0.7) + 40))

            count = (self.h*0.7)+60

            x = self.output_string.split("\n")
            for line in x:
                blitLine = self.Courier_Text_15_Font.render(line, 1, self.codingOutputLetterColor)
                screen.blit(blitLine,((self.w*0.3)+10,count))
                count += 17

            # FPS ticker
            self.clock.tick(self.FPS)
            text = self.HeadlineFont.render("FPS: " + str(round(self.clock.get_fps(),1)),1,(10,10,10))
            screen.blit(text,(0.9 * self.w, 0.9 * self.h))

            ####_________PYGAME__________####
            pygame.display.flip()


    def execute_Code(self, code_string):
        # create file-like string to capture output
        codeOut = StringIO()
        codeErr = StringIO()
        s_err = ""

        # capture output and errors
        sys.stdout = codeOut
        sys.stderr = codeErr

        try:
            exec(code_string)
        # TODO: better error messages (+line in which it occurred)
        except RuntimeError:
            print("A RuntimeError occurred")
            s_err = "RuntimeError"
        except TypeError:
            print("A TypeError occurred!")
            s_err = "TypeError"
        except NameError:
            print("A NameError occurred!")
            s_err = "NameError"
        except ValueError:
            print("A ValueError occurred!")
            s_err = "ValueError"
        except SyntaxError:
            print("SyntaxError: invalid Syntax")
            s_err = "SyntaxError"
        except Exception:
            print("Another unknown error occurred!")
            s_err = "UnknownError"

        # restore stdout and stderr
        sys.stdout = sys.__stdout__
        sys.stderr = sys.__stderr__

        if s_err != "":
            codeOut.close()  # TODO Do I need this in here or can i put it above (duplicate code)
            codeErr.close()
            print("S_ERR")  # TODO Remove
            return (False, s_err)

        s_succ = codeOut.getvalue()
        if s_succ != "":
            codeOut.close() # TODO Do I need this in here or can i put it above (duplicate code)
            codeErr.close()
            print("S_SUCC") # TODO Remove
            return (True, s_succ)

        codeOut.close()
        codeErr.close()
        return (True, "") # Code executed properly, but had no "print" statement.


